rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -57,25 +60,63 @@ class Publisher : public PublisherBase
   Publisher(
     rclcpp::node_interfaces::NodeBaseInterface * node_base,
     const std::string & topic,
-    const rcl_publisher_options_t & publisher_options,
-    const PublisherEventCallbacks & event_callbacks,
-    const std::shared_ptr<MessageAlloc> & allocator)
+    const rclcpp::PublisherOptionsWithAllocator<Alloc> & options,
+    const rclcpp::QoS & qos)
   : PublisherBase(
       node_base,
       topic,
       *rosidl_typesupport_cpp::get_message_type_support_handle<MessageT>(),
-      publisher_options),
-    message_allocator_(allocator)
+      options.template to_rcl_publisher_options<MessageT>(qos)),
+    options_(options),
+    message_allocator_(new MessageAlloc(*options.get_allocator().get()))
```

Comment:
I don't know, maybe I copied it from somewhere else, either way, your suggestion makes sense.

Reply:
So, this is required because the allocator returned by the options is not the same type. They are assignable but not equivalent, so we need to allocate space for the new allocator and then use the conversion constructor to make it.

Really, all the allocators should not be shared pointers, but I'm planning on suggesting that big change in another pull request.